packages/api/src/services/synthetic-prices.ts

```diff
@@ -49,24 +59,101 @@ export default function (config: Config, appState: Dependencies) {
   }
 
-  async function updateLatestPrice(empAddress: string) {
+  // updates raw synth price, in relation to collateral
+  async function updateLatestSynthPrice(empAddress: string) {
     const result: PriceSample = await getSynthPrices.getCurrentPrice(empAddress);
     synthPrices.latest[empAddress] = result;
     return result;
   }
 
+  async function updateLatestSynthPrices(empAddresses: string[]) {
+    // slow down updates by running them serially to prevent rate limits, conform to Promise.allSettled
+    return bluebird.mapSeries(empAddresses, async (empAddress) => {
+      try {
+        return {
+          status: "fulfilled",
+          value: await updateLatestSynthPrice(empAddress),
+        };
+      } catch (err) {
+        return {
+          status: "rejected",
+          reason: err,
+        };
+      }
+    });
+  }
+
+  async function getFullEmpState(empAddress: string) {
+    const emp = await queries.getAnyEmp(empAddress);
+    // the full state has collateral decimals, pulled from erc20 state
+    return queries.getFullEmpState(emp);
+  }
+
+  // gets any price from table, synthetic or collateral. Synthetics go into this table once converted to usd
+  async function getLatestPriceFromTable(empAddress: string, currencyAddress: string) {
+    // PriceSample is type [ timestamp:number, price:string]
+    const priceSample: PriceSample = prices[currency].latest[currencyAddress];
+    assert(uma.utils.exists(priceSample), "No latest price found for emp: " + empAddress);
+
+    const price = priceSample[1];
+    assert(uma.utils.exists(price), "Invalid latest price found on emp: " + empAddress);
+
+    return priceSample;
+  }
+
+  // convert synth price to {currency} which is typically usd, based on the current collateral price
+  async function updateLatestPrice(empAddress: string) {
+    const emp = await queries.getAnyEmp(empAddress);
+
+    assert(uma.utils.exists(emp.collateralCurrency), "Requires contract collateralCurrency: " + empAddress);
+    assert(uma.utils.exists(emp.tokenCurrency), "Requires contract tokenCurrency: " + empAddress);
+
+    const [synthTimestamp, synthPrice] = await getLatestSynthPriceFromTable(empAddress);
+    const [collateralTimestamp, collateralPrice] = await getLatestPriceFromTable(empAddress, emp.collateralCurrency);
+
+    // converted price from raw synth to currency ( usually usd)
+    const price = calcSyntheticPrice(synthPrice, collateralPrice).toString();
+
+    // use the most recent timestamp to index this price
+    const timestamp = Math.max(collateralTimestamp, synthTimestamp);
+    prices[currency].latest[emp.tokenCurrency] = [timestamp, price.toString()];
+    return [timestamp, price];
+  }
+
   async function updateLatestPrices(empAddresses: string[]) {
-    return Promise.allSettled(empAddresses.map(updateLatestPrice));
+    // slow down updates by running them serially to prevent rate limits, conform to Promise.allSettled
```

Comment:
Rather than making each component serialize requests (but only its requests), would it be better to have a global manager for coingecko requests (and any others that are rate sensitive)? The manager can receive inbound requests and ensure a global limit on the number of coingecko requests, for instance.

The reason I suggest this is two fold:
1. Guaranteeing cross-component synchronization. Is there only one place where we talk to coingecko? What happens if the ingestion pipelines starts a second update before the first finishes? All-in-all, it seems tricky to try to put a global rate limit by having each component run the low-level requests that they issue in series unless we're somehow guaranteeing that only one component talks to CG at any one time.
2. This type of throttling depends on the API's response speed. If the API responds quickly, we could go over our per-second quota, which makes this strategy somewhat unpredictable.

Reply:
yea i could see something like this, essentially a wrapper around the coingecko library which automatically rate limits requests.  If the coingecko api returns data on how many requests we have left, that could make it even nicer, as we could throttle calls such that we make them as fast as we can until the reset.  It would be challenging though to generalize this for any type of request.

> Guaranteeing cross-component synchronization. Is there only one place where we talk to coingecko? What happens if the ingestion pipelines starts a second update before the first finishes?  

Currently there is only a single service which depends on coingecko, the `collateral-prices` service. This is by design to avoid some of the problems you are illustrating. There is definitely a disconnect between the interval at which updates happen, and how long they could take, which is something that could be improved.  We can guarantee non overlapping intervals by creating a looping function which begins the timer once all processing is complete.

So what i think i am getting out of this is have a more generalized coingecko rate limiter and be more careful around the frequency of updates to insure they dont overlap. is that right?

Reply:
i ended up adding that async loop function in `apps/api/app.ts`. The idea behind this is that all price updates must finish running before the timer for the next iteration will start.  As for coingecko price rate limit, i looked into this, the api doesnt really return requests remaining or anything like that, but i can make a card for generalizing rate limiting for coingecko and beyond

Reply:
Hmmmm, do you think the top-level looping limit really is necessary if the rate limit problem is fixed? I might argue it's unnecessary complexity as long as we aim to solve the underlying rate limiting issue.

Personally, I think we should add a card, but have no top-level delay. Thoughts?

Reply:
I dont see how we can not have a looping query to coingecko, regardless of if we have native rate limiting.  If the argument is you move the looping to a different spot, that is something we could discuss, though personally i really like controlling all loops at the top level.  lets chat some more on this, it sounds like you have a particular design in mind.